AWSClientRuntime/Sources/Signing/AWSSigV4Signer.swift

```diff
@@ -0,0 +1,68 @@
+//
+// Copyright Amazon.com Inc. or its affiliates.
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import ClientRuntime
+import AwsCommonRuntimeKit
+
+public class AWSSigV4Signer {
+    static let logger: SwiftLogger = SwiftLogger(label: "AWSSigV4Signer")
+    
+    public static func sigV4SignedURL(requestBuilder: SdkHttpRequestBuilder,
+                                      credentialsProvider: CredentialsProvider,
+                                      signingName: Swift.String,
+                                      signingRegion: Swift.String,
+                                      date: ClientRuntime.Date,
+                                      expiration: Int64) -> ClientRuntime.URL? {
+        do {
+            let credentials = try credentialsProvider.getCredentials()
+            
+            let flags = SigningFlags(useDoubleURIEncode: true,
+                                     shouldNormalizeURIPath: true,
+                                     omitSessionToken: false)
+            let signedBodyHeader: AWSSignedBodyHeader = .none
+            let signedBodyValue: AWSSignedBodyValue = .empty
+            let signingConfig = AWSSigningConfig(credentials: credentials,
+                                                 expiration: expiration,
+                                                 signedBodyHeader: signedBodyHeader,
+                                                 signedBodyValue: signedBodyValue,
+                                                 flags: flags,
+                                                 date: date,
+                                                 service: signingName,
+                                                 region: signingRegion,
+                                                 signatureType: .requestQueryParams)
+            let builtRequest = sigV4SignedRequest(requestBuilder: requestBuilder, signingConfig: signingConfig)
+            guard let presignedURL = builtRequest?.endpoint.url else {
+                logger.error("Failed to generate presigend url")
+                return nil
+            }
+            return presignedURL
+        } catch let err {
+            logger.error("Failed to generate presigned url: \(err)")
+            return nil
+        }
+    }
+    
+    public static func sigV4SignedRequest(requestBuilder: SdkHttpRequestBuilder,
+                                          signingConfig: AWSSigningConfig) -> SdkHttpRequest? {
```

Comment:
Is there ever a world in which we would want these apis to just throw instead of returning nil or no?

Reply:
I guess this is more of a philosophical consideration -- when should we throw vs when we should just return `nil`.  I've typically thrown exceptions as a way to convey additional information of some error happening so that the caller can optionally write some code that accounts for that error at the time that the code is running.  For example, if a caller receives a "Busy exception", they might write some code to retry calling the function.

In this case, I could imagine us throwing some sort of "Invalid argument exception" -- however, in this context, we aren't the owners of what is valid/invalid -- we are largely plumbing through values to our downstream dependencies, and I don't think we want to be exposing the exceptions from our lower level dependencies (crt, etc).  We could have a mapping of the CRT errors to exceptions we own, but it seems a bit overkill right now.  What do you think?

A second argument against an exception is that I believe, most of the time, this should just work, so I don't see a strong argument for updating the signature of the function to throw.

Reply:
k this is a valid argument, right now ya we don't have a way to map those errors but we just return them as ClientRuntime.crtError and usually the message in the crt error explains the issue but still not a great experience and I agree it's probably overkill right now to try and improve this because it is still working and functional. Let's leave this for now, I think this is justified.